### PR TITLE
Shibboleth support for library cards

### PR DESCRIPTION
**Pull request by xrosecky**
Shibboleth support for library cards.

TODO: 
- [x] Add a wiki page describing the functionality, including an example configuration

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks, @xrosecky! I will try to give this a review when time permits, though at the moment I am still catching up on some older pull requests due to the holiday backlog. In the meantime, I have also flagged @EreMaijala in case he has any comments, since this touches on a couple of areas that are relevant to his work.

File: themes/bootstrap3/templates/librarycards/editcard.phtml
Comment (EreMaijala) [inline themes/bootstrap3/templates/librarycards/editcard.phtml]: As far as I can see, `$this->shibboleth` check would break normal library card handling if the user logs in using Shibboleth. You should still be able to handle normal library cards regardless of the login method.

Comment (EreMaijala) [reply themes/bootstrap3/templates/librarycards/editcard.phtml]: I'd also rename $rename to $renameOnly.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: Should probably include a check that $id is not 'NEW' if $rename is true. This should be the case, but since the parameters come from the user, they should be validated.

File: themes/bootstrap3/templates/librarycards/home.phtml
Comment (EreMaijala) [inline themes/bootstrap3/templates/librarycards/home.phtml]: This prevents adding of a normal library card if the user has logged in using Shibboleth.

**Review by EreMaijala — CHANGES_REQUESTED** [review]
@xrosecky Could you provide a bit of background information on the use case? Normally, it's thought that either the library card is connected during Shibboleth login, or the user can add their own card later using the card number and pin code.

How does this functionality interact with MyResearch::catalogloginAction that's called when a library card is needed?

I've added some review comments, but they might be a bit off since I'm not sure about why this is needed.

File: themes/bootstrap3/templates/librarycards/editcard.phtml
Comment (xrosecky) [reply themes/bootstrap3/templates/librarycards/editcard.phtml]: It is too specific for our use case so I decided to enable editation of library cards when shibboleth is used for authentication.

**Comment by xrosecky** [comment]
@EreMaijala Can you look at it again? I simplified it and enabled editation of library cards when shibboleth is used.

**Comment by EreMaijala** [comment]
@xrosecky Thanks for the work on simplifying the implementation. I'm still a bit confused as to why this is needed, though. Could you describe the use case in a bit more detail? Normally the library card is connected right at login, and without some further information I'm failing to see when this is needed.

**Comment by xrosecky** [comment]
We need to support connecting multiple library cards to user account using Shibboleth authentication from different IdPs instead of relying on entering card number and pin. In our use case we have separate IdP for each connected library.

**Comment by EreMaijala** [comment]
@xrosecky Ok, now I understand. I'll add some comments when reviewing the code. Have you verified that an existing Shibboleth session from the primary login always gets overridden or is there further configuration required in Shibboleth?

File: module/VuFind/src/VuFind/Auth/Manager.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Auth/Manager.php]: Maybe connectUserCard?

File: module/VuFind/src/VuFind/Auth/Shibboleth.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Auth/Shibboleth.php]: Should probably check that getAttribute returns a valid value here before adding a library card. If not, indicate the error with a flash message or something.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php:256]: Check that getSessionInitiator doesn't return false

File: themes/bootstrap3/templates/librarycards/home.phtml
Comment (EreMaijala) [inline themes/bootstrap3/templates/librarycards/home.phtml]: I think this needs to be configurable and default to being disabled. In most cases the option would probably just confuse users. While adding configuration I'd suggest adding a comment on why and how the feature is used, and if any there are any prerequisites e.g. in Shibboleth configuration.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

**Comment by xrosecky** [comment]
Yes, existing Shibboleth session from the primary login always gets overridden.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: You can combine the two checks into a single `!empty()` check.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Controller/LibraryCardsController.php:256]: While checking for null with the `==` operator works, I'd rather just check `if (!$redirectUrl)` since getSessionInitiator should only return false or a string and the check would cater for both false and an empty string.

File: module/VuFind/src/VuFind/Auth/Manager.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Auth/Manager.php]: This should also be connectUserCard.

File: module/VuFind/src/VuFind/Auth/Shibboleth.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Auth/Shibboleth.php]: This should also be connectUserCard.

File: module/VuFind/src/VuFind/Auth/Shibboleth.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Auth/Shibboleth.php]: I'd prefer `if (!$username)` since it covers all "falsy" cases.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: It would be better to return a redirect with `->toRoute()` to avoid hard-coding paths.

File: module/VuFind/config/module.config.php
Comment (EreMaijala) [inline module/VuFind/config/module.config.php]: The difference between these actions is not obvious from the names. How about naming ConnectNewShibbolethCard to something like ConnectShibbolethCardLogin? Alternatively use a single action and distinquish between the cases using a query parameter, like `/LibraryCards/ConnectShibbolethCard?redirectToLogin=1`.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

**Comment by xrosecky** [comment]
@EreMaijala Could you look at it again?

File: languages/en.ini
Comment (EreMaijala) [inline languages/en.ini]: This just occurred to me: I think it might make sense to change the default wording to align with the actual login option "Institutional Login". I believe that the term "Shibboleth" is something that a user would not typically be aware of.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: Name the view variable something more descriptive like `allowShibbolethCards`.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

File: config/vufind/config.ini
Comment (EreMaijala) [inline config/vufind/config.ini]: I added a TODO point to the task. I'd suggest creating the wiki page and adding a link to it here. Something like "See https://vufind.org/wiki/configuration:shibboleth_library_cards for more information".

Comment (demiankatz) [reply config/vufind/config.ini]: Good idea. Please let me know if you need me to create that page (or increase anyone's wiki rights so that they can do it). (Page creation is a bit restricted right now due to spammers abusing it).

**Comment by demiankatz** [comment]
Thanks for the progress, @xrosecky! Please let me know if you need me to set up a wiki page as discussed above.

**Comment by xmorave2** [comment]
@demiankatz, @xrosecky nor me do not have permissions to create a wiki page, could you please create it? Thanks

**Comment by demiankatz** [comment]
@xmorave2, I have created a placeholder page [here](https://vufind.org/wiki/configuration:shibboleth:library_cards). Please let me know if you have any problems editing it, or if you need anything else. Thanks!

**Comment by xmorave2** [comment]
I wrote a draft of documentation on that page. Any comments are appreciated @demiankatz, @xrosecky. Thanks!

**Comment by demiankatz** [comment]
Thanks, @xmorave2 -- I made some cosmetic changes and adjusted some links so that everything goes somewhere. Feel free to review/revise my work. I decided it was more useful to link directly to config files on GitHub than to create mostly useless wiki pages about those files, but I'm happy to revisit that decision if anyone feels differently. In any case, I'll wait for further feedback before proceeding, but please let me know when everyone is ready for this to be merged!

**Comment by xmorave2** [comment]
Thanks @demiankatz! @xrosecky has reviewed it too. So we are definitely ready for this to be merged.

**Comment by demiankatz** [comment]
Thanks, @xmorave2 -- I think this is nearly ready to go, though I just have one last question: there is a lot of Shibboleth-specific naming still in the code (in the name of the language string, routes, etc.), but the logic actually appears to be fairly generic, as long as the Auth code implements the appropriate method. Should we rename more things with generic names so this does not become confusing if we add support for this feature to a non-Shibboleth auth method? Specifically, I would suggest:

  - Perhaps consider renaming the shibboleth_library_cards setting to auth_based_library_cards (we can still note in the comment that this is currently only supported by Shibboleth, and link to the Shibboleth-specific wiki page).
  - In language files and templates, change not just the value but also the key from "Add a Library Card using Shibboleth login" to "Add a Library Card using institutional login."
  - Remove the word "Shibboleth" from the ConnectShibbolethCard and ConnectShibbolethCardLogin route names; I think this is still appropriate but more concise and generic. (And of course the controller methods should be renamed to match).
  - Rename the allowShibbolethCards variable to something like allowConnectingCards.

I think this makes the code more flexible and future-proof. What do you think?

**Comment by xrosecky** [comment]
Yes, I agree, it is good idea.

**Comment by xrosecky** [comment]
Done - see the latest commit.

File: config/vufind/config.ini
Comment (demiankatz) [inline config/vufind/config.ini:360]: I updated the comment and setting name in this file to match the code; let me know if you have any suggested changes.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: The formatting of this is a bit strange; I'd suggest collapsing this back to a single line, and moving the `&&` to the beginning of the second line to meet the line length restriction.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: I don't think we want this hard-coded Shibboleth restriction anymore. Should we check for the callability of the connectUserCard method here, or just trust that the user configured things correctly and let the system fail elsewhere if there's a problem?

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: Wouldn't it be better to redirect `->toRoute('librarycards-home')` here?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the updates, @xrosecky -- just a few last-minute questions and comments. Once these are addressed, I think we can merge this!

**Comment by xrosecky** [comment]
OK, I fixed them and I also renamed connectUserCard to connectLibraryCard for better naming convention.

File: module/VuFind/src/VuFind/Auth/AbstractBase.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/AbstractBase.php]: How about changing this to `return method_exists($this, 'connectLibraryCard');`? That way, by adding the required method, the supports method will automatically return true, we can remove the override method from the Shibboleth class, and there's less work to implement support for other methods in the future. It's still possible to extend the method if additional platform-specific logic is needed, but it makes the default implementation more flexible.

File: module/VuFind/src/VuFind/Auth/Manager.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/Manager.php]: I'd change `user authenticated by shibboleth` to `authenticated user` to make this more generic.

File: module/VuFind/src/VuFind/Controller/LibraryCardsController.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/LibraryCardsController.php]: Is $config no longer being used? Can we remove this line?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks -- we are very nearly done, but I spotted a couple more minor issues and one potential simplification. I think after this we can definitely merge the PR!

**Comment by xrosecky** [comment]
OK, fixed in the latest commit.

**Comment by demiankatz** [comment]
Thanks, @xrosecky -- merging now!

**Pull request merged**
